src/org/thoughtcrime/securesms/ReactionsHandler.java

```diff
@@ -0,0 +1,151 @@
+import org.thoughtcrime.securesms.database.MessageReactionDatabase;
+import org.thoughtcrime.securesms.database.model.MessageRecord;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class ReactionsHandler {
+    private MessageReactionDatabase reactionsDb;
+
+    public ReactionsHandler(Context context) {
+        this.reactionsDb = DatabaseFactory.getMessageReactionDatabase(context);
+    }
+
+    /**
+     * NOT TEST
+     * This method only used by the conversation fragment since it will automatically
+     * compute the reactorID's to the user itself, Please uses another method if the reaction made
+     * by the recipient. such method can be accessed directly from the MessageReactionDB class
+     * The reaction time will be set to the current system time in milliseconds
+     * @param record
+     * @param reaction
+     */
+    public void reactToMessage(MessageRecord record, String reaction, Long time) {
+        this.reactionsDb.reactToMessage(record, reaction, time);
+    }
+
+    /**
+     * NOT TESTED
+     * This method is ONLY used by JobReceivers since when a reaction received, there
+     * is no active MessageRecord Message object, all what is required the MessageHash being reacted,
+     * reaction, the reactionDate.
+     * This Method can also be used by Fragments however, the developer should handle the time
+     * of the reaction manually
+     * @param messageHash
+     * @param reaction
+     * @param reactionTime
+     * @param reactorID
+     * @param threadId
+     */
+    public void reactToMessage(String messageHash, String reaction, Long reactionTime, Address reactorID, Long threadId) {
+        this.reactionsDb.reactToMessage(messageHash, reaction, reactionTime, reactorID, threadId);
+    }
+
+    public boolean deleteReaction(MessageRecord record) {
+        return this.reactionsDb.removeReaction(record) > 0 ? true:false;
+    }
+
+    public List<Reaction> getMessageReactions(MessageRecord record) {
+        List<Reaction> results = new ArrayList<>();
+        Cursor         cursor  = this.reactionsDb.getMessageReaction(record);
+
+        if (cursor == null) return results;
+
+        while (cursor.moveToNext()) {
+            Parcel c = Parcel.obtain();
+            c.writeString(cursor.getString(cursor.getColumnIndexOrThrow(MessageReactionDatabase.REACTOR_ID)));
+            c.setDataPosition(0);
+            Address reactor = new Address(c);
+
+            Long reactionDate = cursor.getLong(cursor.getColumnIndexOrThrow(MessageReactionDatabase.REACTION_DATE));
+            String reaction = cursor.getString(cursor.getColumnIndexOrThrow(MessageReactionDatabase.REACTION));
+            String reactionId = cursor.getString(cursor.getColumnIndexOrThrow(MessageReactionDatabase.ID));
+
+            String messageType = cursor.getString(cursor.getColumnIndexOrThrow(MessageReactionDatabase.SMS_HASH)) != null
```

Comment:
why are there two columns for mms and sms hash, I thought we had made a decision to not have foreign keys and use a single hash col with a raw string?

Reply:
I thought about it, in case we delete a message, and somehow we end up with dangling hashes, we will not be able to determine if a hash in the reaction table is dangling unless we compare each hash against each table to see if they dangling; however, by differentiating both hashes, I can easily determine which reactions are dangling by simple 1-way join

Reply:
why do we care if hashes are dangling?

Reply:
@g-harel  coz the database will contain reactions to messages that are already deleted, this will degrade the query performance in the long term. we have a method deletes all those dangling reaction"hashes" that we can run on app start